test/e2e/simulateTrade.test.ts

```diff
@@ -0,0 +1,297 @@
+    const wethUsdt = await createPair(weth.address, usdt.address);
+    await weth.mint(wethUsdt.address, ethers.utils.parseEther("1000.0"));
+    await usdt.mint(wethUsdt.address, ethers.utils.parseUnits("4500000.0", 6));
+    await wethUsdt.mint(pooler.address);
+
+    await weth.mint(trader.address, ethers.utils.parseEther("1.000"));
+    await weth
+      .connect(trader)
+      .approve(vaultRelayer.address, ethers.constants.MaxUint256);
+
+    const swapInteractions = [
+      {
+        target: weth.address,
+        callData: weth.interface.encodeFunctionData("approve", [
+          uniswapRouter.address,
+          ethers.constants.MaxUint256,
+        ]),
+      },
+      {
+        target: uniswapRouter.address,
+        callData: uniswapRouter.interface.encodeFunctionData(
+          "swapExactTokensForTokens",
+          [
+            ethers.utils.parseEther("0.999"),
+            1, // market order, YOLO!
+            [weth.address, usdt.address],
+            settlement.address,
+            0xffffffff,
+          ],
+        ),
+      },
+    ];
+
+    const { gasUsed: simulatedGasUsed, executedBuyAmount } =
+      await tradeSimulator.simulateTrade(
+        {
+          owner: trader.address,
+          sellToken: weth.address,
+          buyToken: usdt.address,
+          sellAmount: ethers.utils.parseEther("0.999"),
+          buyAmount: 0,
+        },
+        {
+          [InteractionStage.INTRA]: swapInteractions,
+        },
+      );
+
+    const encoder = new SettlementEncoder(domainSeparator);
+    await encoder.signEncodeTrade(
+      {
+        kind: OrderKind.SELL,
+        partiallyFillable: false,
+        sellToken: weth.address,
+        buyToken: usdt.address,
+        sellAmount: ethers.utils.parseEther("0.999"),
+        buyAmount: ethers.utils.parseUnits("4000.0", 6),
+        feeAmount: ethers.utils.parseEther("0.001"),
+        validTo: 0xffffffff,
+        appData: 1,
+      },
+      trader,
+      SigningScheme.EIP712,
+    );
+    for (const interaction of swapInteractions) {
+      encoder.encodeInteraction(interaction);
+    }
+
+    const tx = await settlement.connect(solver).settle(
+      ...encoder.encodedSettlement({
+        [weth.address]: executedBuyAmount,
+        [usdt.address]: ethers.utils.parseEther("0.999"),
+      }),
+    );
+    const { transactionHash, gasUsed: actualGasUsed } = await tx.wait();
+
+    expect(await weth.balanceOf(trader.address)).to.equal(0);
+    expect(await weth.balanceOf(settlement.address)).to.equal(
+      ethers.utils.parseEther("0.001"),
+    );
+    expect(await usdt.balanceOf(trader.address)).to.equal(executedBuyAmount);
+
+    const gasOverhead = await transactionGasOverhead(transactionHash);
+    const gasUsageRatio =
+      (actualGasUsed.toNumber() - simulatedGasUsed.toNumber() - gasOverhead) /
+      actualGasUsed.toNumber();
+    console.log(
+      `Simulated trade used ${simulatedGasUsed} (+${gasOverhead}) and cost ${actualGasUsed}`,
```

Comment:
I'd like to see the gas difference between the two printed out, by doing the math it seems to be about 12k, so fairly close to 5%?